File: 2022/07/15/java/ReverseString.java
/**
 * https://leetcode.com/problems/reverse-string/
 * @author Andrew Jarombek
 * @since 7/15/2022
 */

public class ReverseString {
    public static void reverseString(char[] s) {
        int start = 0;
        int end = s.length - 1;

        while (start < end) {
            char c = s[start];
            s[start] = s[end];
            s[end] = c;

            start++;
            end--;
        }
    }

    // Less memory usage, but longer runtime.
    // Solution #1 is likely preferable due to the faster runtime.
    public static void reverseStringV2(char[] s) {
        int start = 0;

        while (start < s.length / 2) {
            char c = s[s.length - start - 1];
            s[s.length - start - 1] = s[start];
            s[start] = c;

            start++;
        }
    }

    public static void main(String[] args) {
        char[] chars = new char[] {'H', 'e', 'l', 'l', 'o'};
        reverseString(chars);

        assert chars[0] == 'o';
        assert chars[1] == 'l';
        assert chars[2] == 'l';
        assert chars[3] == 'e';
        assert chars[4] == 'H';

        reverseStringV2(chars);

        assert chars[0] == 'H';
        assert chars[1] == 'e';
        assert chars[2] == 'l';
        assert chars[3] == 'l';
        assert chars[4] == 'o';
    }
}

File: 2022/09/10/reverseString.js
/**
 * https://leetcode.com/problems/reverse-string-ii/
 * @author Andrew Jarombek
 * @since 9/10/2022
 */

const reverseStr = function(s, k) {
    let index = 0;
    let result = [];

    while (index < s.length) {
        let end = Math.min(s.length - 1, index + k - 1);
        for (let i = end; i >= index; i--) {
            result.push(s.charAt(i));
        }

        index = end + 1;
        end = Math.min(s.length, index + k);

        for (let i = index; i < end; i++) {
            result.push(s.charAt(i));
        }

        index = end;
    }

    return result.join("");
};

console.assert(reverseStr("abcdefgh", 4) === "dcbaefgh");
console.assert(reverseStr("abcd", 2) === "bacd");
